### ISSUE TITLE
Alternation and Concatenation rules cause stack overflow

### ISSUE THREAD

**Issue by RobertWalter83** [manuscrypt/elm-ebnf#1]
It seems like the functions "alternation" and "concatenation" cause an endless recursion when typing something like this in the editor:

`g =`

I first assumed that the end of the file is not recognized correctly, but I am not so sure this is true since the problem also occurs if you start out with somehting like this:

`r;`

and then move the cursor to the front and type `g =` again. 

In any case, if you comment out the calls to those rules, the stack overflow does not happen.

```
rhs : List (Parser Rhs)
rhs =
    [ --lazy (\_ -> alternation)
    --, lazy (\_ -> concatenation)
      lazy (\_ -> option)
    , lazy (\_ -> grouping)
    , lazy (\_ -> repetition)
    , terminal
    , ident
    ]

concatenation : Parser Rhs
concatenation =
    succeed Concatenation
        |= oneOf rhs
        |= con


alternation : Parser Rhs
alternation =
    succeed Alternation
        |= oneOf rhs
        |= alt
```
I assume that their "oneOf rhs" call causes that. Nothing is consumed, maybe, so the same rule gets applied over and over again? Just a guess, since I still struggle with the syntax of the parser package a bit.

**Comment by manuscrypt** [comment]
Yes, when I remove those, it works as expected, but it's not very useful. I
will ask on slack again later.

**Comment by RobertWalter83** [comment]
Well, removing the calls was of course not a proposed solution ;)
I think the problem is that with this code, you produce a indirect self recursion:

```
rhs : List (Parser Rhs)
rhs =
    [ lazy (\_ -> alternation)
     ... ]

concatenation : Parser Rhs
concatenation =
    succeed Concatenation
        |= oneOf rhs
        |= con
```

It's like saying (in EBNF-like):

```
A := B | C | D ;
B := A
``` 
Not sure how to solve it, unfortunately (so maybe that's what you want to ask), but I wanted to share what I was able to diagnose :)

**Comment by manuscrypt** [comment]
Thank you for your comments. I feel what you are suggesting is an additional problem that would occur when trying to generate source code from the parsed grammar (need to detect cycles). Currently, the parser should not give any "meaning" to the parsed parts. I still want to believe it's solvable :)
